File: frontend_web/src/services/api/employeePlanning.ts
import api from './api';

export interface EmployeePlanningData {
    id?: number;
    employee_id: number;
    weekday: string;
    available: boolean;
    custom_text?: string;
    replacement_id?: number;
    replacement_employee?: {
        id: number;
        first_name: string;
        last_name: string;
        function: string;
    };
    calendar_week?: number;
    created_at?: string;
    updated_at?: string;
    has_conflicts?: boolean;
    appointments_count?: number;
    patient_count?: number;
    replacement_affected_count?: number;
}


export const employeePlanningApi = {
    // Get all planning entries for current week
    getAll: (calendarWeek?: number) => {
        const params = calendarWeek ? `?calendar_week=${calendarWeek}` : '';
        return api.get(`/employee-planning/${params}`);
    },


    // Update replacement for employee and weekday
    updateReplacement: (employeeId: number, weekday: string, data: {
        replacement_id?: number;
        calendar_week?: number;
    }) => {
        return api.put(`/employee-planning/${employeeId}/${weekday}/replacement`, data);
    },

    

};

File: backend/app/models/employee_planning.py
from datetime import datetime

from app import db


class EmployeePlanning(db.Model):
    __tablename__ = "employee_planning"

    id = db.Column(db.Integer, primary_key=True)
    employee_id = db.Column(db.Integer, db.ForeignKey("employees.id"), nullable=False)
    weekday = db.Column(db.String(20), nullable=False)  # monday, tuesday, etc.
    available = db.Column(db.Boolean, nullable=False, default=True)
    custom_text = db.Column(db.String(200), nullable=True)  # For custom status
    replacement_id = db.Column(
        db.Integer, db.ForeignKey("employees.id"), nullable=True
    )  # Replacement employee
    calendar_week = db.Column(db.Integer, nullable=True)  # Denormalized for easier filtering
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    employee = db.relationship("Employee", foreign_keys=[employee_id], backref="planning_entries")
    replacement = db.relationship(
        "Employee", foreign_keys=[replacement_id], backref="replacement_entries"
    )

    # Unique constraint: one planning entry per employee per weekday per week
    __table_args__ = (
        db.UniqueConstraint(
            "employee_id", "weekday", "calendar_week", name="unique_employee_weekday_week"
        ),
    )

    def to_dict(self):
        return {
            "id": self.id,
            "employee_id": self.employee_id,
            "weekday": self.weekday,
            "available": self.available,
            "custom_text": self.custom_text,
            "replacement_id": self.replacement_id,
            "calendar_week": self.calendar_week,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }

    @staticmethod
    def from_dict(data):
        return EmployeePlanning(
            employee_id=data.get("employee_id"),
            weekday=data.get("weekday"),
            available=data.get("available", True),
            custom_text=data.get("custom_text"),
            replacement_id=data.get("replacement_id"),
            calendar_week=data.get("calendar_week"),
        )
